### PR TITLE
Optimization library

### PR DESCRIPTION
**Pull request by breuleux**
This adds a set of optimizers:
* Tuple-related optimizations like `(x, y, ...)[0] => x`.
* A few trivial arithmetic simplifications.
* Constant propagation.
* Inlining.

This also implements a few small fixes:
* Bug in `GraphCloner` when there are unused parameters in cloned graphs.
* Stack overflow in isomorphic on recursive functions.

### CODE REVIEW

**Comment by breuleux** [comment]
I retired GraphUnification now-ish.

File: myia/ir/anf.py [hunk 1]
```diff
@@ -322,6 +339,9 @@ def __init__(self, value: Any) -> None:
         """Construct a literal."""
         super().__init__([], value, None)
 
+    def __visit__(self, fn):
+        return Constant(fn(self.value))
```
Comment (abergeron) [inline myia/ir/anf.py]: noseq

File: myia/ir/anf.py [hunk 2]
```diff
@@ -282,6 +283,11 @@ def __init__(self, inputs: List[ANFNode], graph: 'Graph') -> None:
         """Construct an application."""
         super().__init__(inputs, APPLY, graph)
 
+    def __visit__(self, fn):
+        new_inputs = expandlist(map(fn, self.inputs))
+        g = fn(self.graph)
```
Comment (abergeron) [inline myia/ir/anf.py]: you probably want noseq here

File: myia/ir/anf.py [hunk 3]
```diff
@@ -300,6 +306,17 @@ def __init__(self, graph: Graph) -> None:
         """Construct the parameter."""
         super().__init__([], PARAMETER, graph)
 
+    def __visit__(self, fn):
+        g = fn(self.graph)
```
Comment (abergeron) [inline myia/ir/anf.py]: noseq

**Pull request merged**